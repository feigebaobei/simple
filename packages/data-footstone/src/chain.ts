import {
  BaseChainElement as BCE,
  BaseChain as BC,
  SingleChain as SC,
  SingleChainElement as SCE,
  DoublyChain as DC,
  DoublyChainElement as DCE,
  SingleCircleChain as SCC,
  SingleCircleChainElement as SCCE,
  DoublyCircleChainElement as DCCE,
  DoublyCircleChain as DCC,
  N,
  B,
} from '../typings'

class BaseChain<T> implements BC<T> {
  head: BCE<T>
  length: N
  constructor() {
    this.head = null
    this.length = 0
  }
  toArray() {
    let cur = this.head
    let arr = []
    while (cur) {
      arr.push(cur.value)
      cur = cur.next
    }
    return arr
  }
  // 默认至少有一个节点
  isValidRange(p: N) {
    return 0 <= p && p < this.length
  }
  // 返回匹配元素的下标
  indexOf(v: T, all = false) {
    if (this.length) {
      let index = 0
      let cur = this.head
      let res: N[] | N // = []
      if (all) {
        res = []
        while (cur) {
          if (cur.value === v) {
            res.push(index)
          }
          cur = cur.next
          index++
        }
      } else {
        res = -1
        while (cur) {
          if (cur.value === v) {
            res = index
            break
          }
          cur = cur.next
          index++
        }
      }
      return res
    } else {
      return -1
    }
  }
  getEleByIndex(i: N) {
    if (this.isValidRange(i)) {
      let index = 0
      let cur = this.head
      let res: T | undefined
      while (cur) {
        if (index === i) {
          res = cur.value
        }
        cur = cur.next
        index++
      }
      return res
    } else {
      return undefined
    }
  }
}

class SingleChain<T> extends BaseChain<T> implements SC<T> {
  head: SCE<T>
  length: N
  constructor(...p: T[]) {
    super()
    if (p.length) {
      this.head = p.reduceRight((r: SCE<T>, c: T, index: N) => {
        return {
          value: c,
          next: r,
          position: index,
        }
      }, null)
      this.length = p.length
    }
  }
  createNode(v: T, p: N) {
    return {
      value: v,
      next: null,
      position: p,
    }
  }
  // 对外不暴露数据结构。所心不提供appendNode方法。
  append(p: T) {
    let node = this.createNode(p, this.length)
    if (!this.head) {
      this.head = node
    } else {
      let cur = this.head
      while (cur.next) {
        cur = cur.next
      }
      cur.next = node
    }
    this.length++
    return this.length
  }
  // 不允许插入到最后一个
  insert(v: T, p: N) {
    if (this.isValidRange(p)) {
      let res: B
      let node = this.createNode(v, p)
      if (p === 0) {
        node.next = this.head
        this.head = node
      } else {
        let index = 0
        let cur = this.head
        let pre = null
        while (index < p) {
          pre = cur
          cur = cur.next
          index++
        }
        pre.next = node
        node.next = cur
      }
      this.setPosition(p)
      this.length++
      return res
    } else {
      return false
    }
  }
  // 返回被删除的元素或undefined
  removeAt(p: N) {
    if (this.isValidRange(p)) {
      let res: T
      if (p === 0) {
        res = this.head.value
        this.head = this.head.next
      } else {
        let index = 0
        let cur = this.head
        let pre = null
        while (index < p) {
          pre = cur
          cur = cur.next
          index++
        }
        pre.next = cur.next
        res = cur.value
      }
      this.setPosition(p)
      this.length--
      return res
    } else {
      return undefined
    }
  }
  // 删除此方法
  // removeElement(v: T, all = false) {
  //   let res: B = false
  //   if (this.length) {
  //     if (this.head.value === v) {
  //       this.head = this.head.next
  //       this.length--
  //       res = true
  //     } else {
  //       let cur = this.head.next
  //       let pre = this.head
  //       while (cur) {
  //         if (cur.value === v) {
  //           pre.next = cur.next
  //           this.length--
  //           res = true
  //           if (!all) {
  //             break
  //           }
  //         }
  //         pre = cur
  //         cur = cur.next
  //       }
  //     }
  //     let cur = this.head
  //     let index = 0
  //     while (cur) {
  //       cur.position = index
  //       index++
  //       cur = cur.next
  //     }
  //   } else {
  //     res = false
  //   }
  //   return res
  // }
  // includes() {}
  reverseSelf() {
    let fn = (p: SCE<T> | null, q = null) => {
      if (p) {
        return fn(p.next, { value: p.value, next: q })
      } else {
        return q
      }
    }
    this.head = fn(this.head)
    this.setPosition()
    return this
  }
  reverse() {
    return this.toArray().reduceRight((r, c) => {
      r.append(c)
      return r
    }, new SingleChain())
  }
  clear() {
    this.head = null
    this.length = 0
  }
  setPosition(from: N = 0) {
    let cur = this.head
    let index = 0
    while (cur) {
      if (index >= from) {
        cur.position = index
      }
      index++
      cur = cur.next
    }
  }
}

class DoublyChain<T> extends BaseChain<T> implements DC<T> {
  head: DCE<T> | null
  tail: DCE<T> | null
  length: N
  constructor(...p: T[]) {
    super()
    this.tail = null
    p.forEach(this.append, this)
  }
  createNode(v: T, p: N) {
    return {
      value: v,
      position: p,
      prev: null,
      next: null,
    }
  }
  append(v: T) {
    let node = this.createNode(v, this.length)
    if (this.length) {
      node.prev = this.tail
      this.tail.next = node
      this.tail = node
    } else {
      this.head = node
      this.tail = node
    }
    this.length++
    return this.length
  }
  insert(v: T, p: N) {
    let res = false
    if (this.isValidRange(p)) {
      let node = this.createNode(v, p)
      if (p === 0) {
        node.next = this.head
        this.head.prev = node
        this.head = node
      } else {
        let cur = this.head
        let index = 0
        while (index < p) {
          cur = cur.next
          index++
        }
        cur.prev.next = node
        node.prev = cur.prev
        node.next = cur
        cur.prev = node
      }
      this.setPosition(p)
      this.length++
      res = true
    } else {
      res = false
    }
    return res
  }
  // 返回被删除的元素
  removeAt(p: N) {
    if (this.isValidRange(p)) {
      let res: T
      if (p === 0) {
        res = this.head.value
        if (this.length === 1) {
          this.head = null
          this.tail = null
        } else {
          this.head = this.head.next
          this.head.prev = null
        }
      } else if (p === this.length - 1) {
        res = this.tail.value
        this.tail = this.tail.prev
        this.tail.next = null
      } else {
        let cur = this.head
        let index = 0
        while (index < p) {
          cur = cur.next
          index++
        }
        res = cur.value
        cur.prev.next = cur.next
        cur.next.prev = cur.prev
      }
      this.setPosition(p)
      this.length--
      return res
    } else {
      return undefined
    }
  }
  // removeElement(v: T, all = false) {
  //   let res: B = false
  //   let cur = this.head
  //   let index = 0
  //   while (index < this.length) {
  //     if (cur.value === v) {
  //       if (index === 0) {
  //         // op this.head
  //         this.head = this.head.next
  //         this.head.prev = null
  //       } else if (index === this.length) {
  //         // op this.tail
  //         this.tail = this.tail.prev
  //         this.tail.next = null
  //       } else {
  //         // op middle
  //         cur.prev.next = cur.next
  //         cur.next.prev = cur.prev
  //       }
  //       res = true
  //       this.length--
  //       if (!all) {
  //         break
  //       }
  //     } else {
  //       index++
  //     }
  //     cur = cur.next
  //   }
  //   return res
  // }
  clear() {
    this.head = null
    this.tail = null
    this.length = 0
  }
  setPosition(from: N = 0) {
    let cur = this.head
    let index = 0
    while (cur) {
      if (index >= from) {
        cur.position = index
      }
      index++
      cur = cur.next
    }
  }
}

class SingleCircleChain<T> extends SingleChain<T> implements SCC<T> {
  head: SCCE<T> | null
  length: N
  tail: SCCE<T> | null
  constructor(...p: T[]) {
    super()
    this.head = null
    this.tail = null
    this.length = 0
    p.forEach(this.append, this)
  }
  toArray() {
    let res = []
    let index = 0
    let cur = this.head
    while (index < this.length) {
      res.push(cur.value)
      cur = cur.next
      index++
    }
    return res
  }
  createNode(v: T, p: N): SCCE<T> {
    return {
      value: v,
      position: p,
      next: null,
    }
  }
  append(v: T) {
    let node = this.createNode(v, this.length)
    if (this.length) {
      this.tail.next = node
      node.next = this.head
      this.tail = node
    } else {
      this.head = node
      this.tail = node
      node.next = node
    }
    this.length++
    return this.length
  }
  // 允许的范围  [0, length)
  insert(v: T, p: N) {
    if (this.isValidRange(p)) {
      let node = this.createNode(v, p)
      if (p === 0) {
        this.tail.next = node
        node.next = this.head
        this.head = node
      } else {
        let index = 0
        let cur = this.head
        let pre = null
        while (index < p) {
          pre = cur
          cur = cur.next
          index++
        }
        pre.next = node
        node.next = cur
      }
      this.length++
      this.setPosition(p)
      return true
    } else {
      return false
    }
  }
  // 删除指定位置的节点，返回被删除的元素或undefined
  removeAt(p: N) {
    if (this.isValidRange(p)) {
      let res: T
      if (p === 0) {
        res = this.head.value
        this.head = this.head.next
        this.tail.next = this.head
      } else {
        let index = 0
        let cur = this.head
        let pre = null
        while (index < p) {
          pre = cur
          cur = cur.next
          index++
        }
        pre.next = cur.next
      }
      this.length--
      this.setPosition(p)
      return res
    } else {
      return undefined
    }
  }
  setPosition(from: N = 0) {
    let index = 0
    let cur = this.head
    while (index < this.length) {
      if (index >= from) {
        cur.position = index
      }
      index++
      cur = cur.next
    }
  }
  // 返回 boolean 表示是否删除
  // removeElement(v: T, all = false) {
  //   let res: B = false
  //   if (this.length) {
  //     if ((this.head.value = v)) {
  //       this.head.next.prev = this.head.prev
  //       this.head.prev.next = this.head.next
  //       this.length--
  //       res = true
  //     } else {
  //       let cur = this.head.next
  //       while (cur) {
  //         if (cur.value === v) {
  //           cur.prev.next = cur.next
  //           cur.next.prev = cur.prev
  //           this.length--
  //           res = true
  //           if (!all) {
  //             break
  //           }
  //         }
  //         cur = cur.next
  //       }
  //     }
  //   } else {
  //     res = false
  //   }
  //   return res
  // }
}

class DoublyCircleChain<T> extends DoublyChain<T> implements DCC<T> {
  constructor(...p: T[]) {
    super()
    p.forEach(this.append, this)
  }
  createNode(v: T, p: number) {
    return {
      value: v,
      position: p,
      next: null,
      prev: null,
    }
  }
  append(p: T) {
    let node = this.createNode(p, this.length)
    if (this.length) {
      node.next = this.head
      this.head.prev = node
      this.tail.next = node
      node.prev = this.tail
      this.tail = node
    } else {
      node.prev = node
      node.next = node
      this.head = node
      this.tail = node
    }
    this.length++
    return this.length
  }
  insert(v: T, p: N) {
    let res = false
    if (this.isValidRange(p)) {
      let node = this.createNode(v, p)
      if (p === 0) {
        node.prev = node
        node.next = node
        this.head = node
        this.tail = node
      } else {
        let cur = this.head
        let index = 0
        while (index < p) {
          cur = cur.next
          index++
        }
        cur.prev.next = node
        node.prev = cur.prev
        node.next = cur
        cur.prev = node
      }
      this.length++
      this.setPosition(p)
      res = true
    } else {
      res = false
    }
    return res
  }
  removeAt(p: N) {
    if (this.isValidRange(p)) {
      let res: T
      if (p === 0) {
        res = this.head.value
        if (this.length === 1) {
          this.head = null
          this.tail = null
        } else {
          this.head = this.head.next
          this.tail.next = this.head
          this.head.prev = this.tail
        }
      } else if (p === this.length - 1) {
        res = this.tail.value
        this.tail = this.tail.prev
        this.tail.next = this.head
      } else {
        let cur = this.head
        let index = 0
        while (index < p) {
          cur = cur.next
          index++
        }
        res = cur.value
        cur.prev.next = cur.next
        cur.next.prev = cur.prev
      }
      this.setPosition(p)
      this.length--
      return res
    } else {
      return undefined
    }
  }
  // removeElement(v: T, all = false) {
  //   let res: B = false
  //   let cur = this.head
  //   let index = 0
  //   while (index < this.length) {
  //     if (cur.value === v) {
  //       if (index === 0) {
  //         this.head = this.head.next
  //         this.head.prev = this.tail
  //         this.tail.next = this.head
  //       } else if (index === this.length) {
  //         this.tail = this.tail.prev
  //         this.tail.next = this.head
  //         this.head.prev = this.tail
  //       } else {
  //         cur.prev.next = cur.next
  //         cur.next.prev = cur.prev
  //       }
  //       res = true
  //       this.length--
  //       if (!all) {
  //         break
  //       }
  //     } else {
  //       index++
  //     }
  //     cur = cur.next
  //   }
  //   return res
  // }
  toArray() {
    let res = []
    let index = 0
    let cur = this.head
    while (index < this.length) {
      res.push(cur.value)
      index++
      cur = cur.next
    }
    return res
  }
  setPosition(from: N = 0) {
    let index = 0
    let cur = this.head
    while (index < this.length) {
      if (index >= from) {
        cur.position = index
      }
      index++
      cur = cur.next
    }
  }
}

export { SingleChain, DoublyChain, SingleCircleChain, DoublyCircleChain }
