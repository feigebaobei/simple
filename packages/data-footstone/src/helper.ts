import {
     F, N, S, A
  } from './typings'
  
  
let af = Array.from
let NPF = Number.POSITIVE_INFINITY
function mapToObj (strMap: Map<S, A>) {
    let obj = Object.create(null)
    for (let [key, value] of strMap) {
        obj[key] = value
    }
    return obj
}
export {af, NPF, mapToObj}