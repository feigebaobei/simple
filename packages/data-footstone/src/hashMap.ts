import { SingleChain } from './chain'
import {
  HashMap as HM,
  HashMapKind as HMK,
  HashMapHash as HMH,
  HashFn,
  N,
  S,
  A,
} from '../typings'

let djb2HashFn: HashFn = (k: A) => {
  let h = 5381
  let t = String(k)
  for (let i = 0; i < t.length; i++) {
    h = h * 33 + t.charCodeAt(i)
  }
  return h % 1013
}
let loseloseHashFn: HashFn = (k: A) => {
  let h = 0
  let t = String(k)
  for (let i = 0; i < t.length; i++) {
    h = h + t.charCodeAt(i)
  }
  return h % 37
}

class HashMap<G> implements HM<G> {
  box: HM<G>['box'] // 日后都改为这种写法
  _size: N
  kind: HMK
  hash: HMH
  _put: HM<G>['_put']
  _hashFn: (k: A) => N
  // _get: (k: A) => N
  _get: HM<G>['_get']
  _remove: (k: A) => G
  constructor(kind: HMK = 'separate', hash: HMH = 'djb2') {
    this.box = []
    this._size = 0
    this.kind = kind
    this.hash = hash
    switch (kind) {
      case 'separate':
      default:
        this._put = (k: A, v: G) => {
          let node = this._get(k)
          if (node) {
            node.value = v
          } else {
            let p = this.hashFn(k)
            if (this.box[p] === undefined) {
              this.box[p] = new SingleChain()
            }
            this.box[p].append(this.createNode(k, v))
            this._size++
          }
          return this._size
        }
        this._get = (k) => {
          let p = this.hashFn(k)
          if (this.box[p] === undefined) {
            return undefined
          } else {
            let cur = this.box[p].head
            let res = undefined
            while (cur) {
              if (cur.value.key === k) {
                // res = cur.value.value
                // 改为返回节点
                res = cur.value
                break
              }
              cur = cur.next
            }
            return res
          }
        }
        this._remove = (k) => {
          let p = this.hashFn(k)
          let res = undefined
          if (this.box[p] === undefined) {
            res = undefined
          } else {
            let cur = this.box[p].head
            let index = 0
            while (cur) {
              if (cur.value.key === k) {
                res = cur.value.value
                this.box[p].removeAt(index)
                break
              }
              index++
            }
          }
          this._size--
          return res
        }
        break
      case 'line':
        this._put = (k: S, v: G) => {
          let node = this._get(k)
          if (node) {
            node.value = v
          } else {
            let p = this.hashFn(k)
            while (this.box[p] !== undefined) {
              p++
            }
            this.box[p] = this.createNode(k, v)
            this._size++
          }
          return this._size
        }
        this._get = (k) => {
          let p = this.hashFn(k)
          let res = undefined
          while (this.box[p] !== undefined) {
            if (this.box[p].key === k) {
              // res = this.box[p].value
              // 改为返回节点
              res = this.box[p]
              break
            }
            p++
          }
          return res
        }
        this._remove = (k) => {
          let p = this.hashFn(k)
          let res = undefined
          while (this.box[p] !== undefined) {
            if (this.box[p].key === k) {
              res = this.box[p].value
              this.box[p] = undefined
              break
            }
            p++
          }
          this._size--
          return res
        }
        break
    }
    switch (hash) {
      case 'djb2':
      default:
        this._hashFn = djb2HashFn
        break
      case 'loselose':
        this._hashFn = loseloseHashFn
        break
    }
  }
  createNode(k: A, v: G) {
    return {
      key: k,
      value: v,
    }
  }
  put(k: A, v: G) {
    return this._put(k, v)
  }
  get(k: A) {
    let res = this._get(k)
    return res ? res.value : undefined
  }
  remove(k: A) {
    return this._remove(k)
  }
  size() {
    return this._size
  }
  hashFn(k: A) {
    return this._hashFn(k)
  }
}

// 也可以做成：
// 在基类中定义_hashFn等方法
// 在扩展类中使用基类的方法

export { HashMap, djb2HashFn, loseloseHashFn }
