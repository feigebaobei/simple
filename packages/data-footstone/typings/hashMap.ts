import { N, A } from './baseType'
import { SingleChain } from './chain'

type HashMapKind = 'separate' | 'line'
type HashMapHash = 'djb2' | 'loselose'
interface HashMapBoxItem<G> {
  key: A
  value: G
}
type HashMapBox<G> = HashMapBoxItem<G>[]

type HashFn = (p: A) => N

interface HashMap<G> {
  box: SingleChain<G> | HashMapBox<G>
  _size: N
  kind: HashMapKind
  hash: HashMapHash
  createNode: (k: A, v: G) => HashMapBoxItem<G>
  _put: (k: A, v: G) => N
  _hashFn: (k: A) => N
  _remove: (k: A) => G
  _get: (k: A) => HashMapBoxItem<G> | undefined
  put: (k: A, v: G) => N
  remove: (k: A) => G
  get: (k: A) => G | undefined
  hashFn: (k: A) => N
  size: () => N
}

export { HashMapKind, HashMapHash, HashMap, HashFn }
