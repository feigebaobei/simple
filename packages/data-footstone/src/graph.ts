import {
  Vertex as V,
  Edge as E,
   Graph as G,
   Queue as Q, 
  GraphColor as GC,
   F, N, B 
} from '../typings'
import { Queue } from './queue'
import { af } from './helper'

class Graph<T> implements G<T> {
  vertexMap: G<T>['vertexMap']
  adjMatrix: G<T>['adjMatrix']
  _adjTable: G<T>['_adjTable']
  direction: G<T>['direction']
  constructor (direction: B = true) {
    this.vertexMap = new Map()
    // this.edgeMap = new Map()
    this.adjMatrix = new Map() // 保存边的信息
    this._adjTable = new Map()
    // this.direction = direction // 是否有方向
    // 不可改变
    Object.defineProperty(this, 'direction', {
      value: direction,
      writable: false
    })
  }
  createVertex(v: T) {
    return {
      data: v,
      inDegree: 0,
      outDegree: 0,
      status: '',
      dTime: new Date(),
      fTime: new Date(),
    }
  }
  createEdge(a: T, b: T) {
    return {
      data: '',
      start: this.vertexMap.get(a),
      end: this.vertexMap.get(b),
      weight: 0,
      status: ''
    }
  }
  get adjTable () {
    return af(this._adjTable.entries()).reduce((r, [k, v]) => {
      r.set(k, af(v.keys()).reduce((r, c) => {
        this.vertexMap.get(c)
        return r
      }, new Map()))
      return r
    }, new Map)
  }
  putVertex(v: T) {
    this.vertexMap.set(v, this.createVertex(v))
    for (let a of this.adjMatrix.values()) {
      a.set(v, null)
    }
    this.adjMatrix.set(v, new Map(Array.from(this.vertexMap.keys()).map((v) => [v, null])))
    this._adjTable.set(v, new Set())
  }
  putEdge(a: T, b: T) {
    if (this.direction) { // 有方向，需要添加1个边
      this.adjMatrix.get(a).set(b, this.createEdge(a, b))
      this._adjTable.get(a).add(b)
    } else { // 无方向，需要添加2个边。
      this.adjMatrix.get(a).set(b, this.createEdge(a, b)) // 添加 a->b
      this.adjMatrix.get(b).set(a, this.createEdge(b, a)) // 添加 b->a
      this._adjTable.get(a).add(b)
      this._adjTable.get(b).add(a)
    }
  }
  edgeList() {
    return Array.from(this.adjMatrix.values()).reduce((r: E<T>[], c: Map<T, E<T>>) => {
      r.push(...Array.from(c.values()))
      return r
    }, []).filter(Boolean, this)
  }
  removeVertex(a: T) {
    let res = this.vertexMap.get(a)
    if (res) {
      this.vertexMap.delete(a) // 删除点
      this.adjMatrix.delete(a)
      af(this.adjMatrix.values()).forEach(map => map.delete(a))
      this._adjTable.delete(a)
      af(this._adjTable.values()).forEach(set => set.delete(a))
    }
    return res
  }
  removeEdge(a: T, b: T) {
    let res = []
    if (this.direction) { // 有方向，应该删除1个边
      let t = this.adjMatrix.get(a)
      res.push(t.get(b))
      t.delete(b)
      this._adjTable.get(a).delete(b)
    } else { // 无方向，应该删除2个边
      let t = this.adjMatrix.get(a)
      res.push(t.get(b))
      t.delete(b)
      t = this.adjMatrix.get(b)
      res.push(t.get(a))
      t.delete(a)
      this._adjTable.get(a).delete(b)
      this._adjTable.get(b).delete(a)
    }
    return res
  }




  // vertices: G<T>['vertices']
  // adjList: G<T>['adjList']
  // // adjMatrix: G<T>['adjMatrix']
  // // static vertices: any
  // constructor() {
  //   this.vertices = []
  //   this.adjList = new Map()
  //   // this.adjMatrix = []
  // }
  // putVertex(v: T) {
  //   this.vertices.push(v)
  //   this.adjList.set(v, [])
  // }
  // putEdge(v: T, w: T) {
  //   this.adjList.get(v).push(w)
  //   this.adjList.get(w).push(v)
  // }
  // initColor() {
  //   let color: Map<T, GC> = new Map()
  //   this.vertices.forEach((item) => {
  //     color.set(item, 'white')
  //   })
  //   return color
  // }
  // bfs(index = 0, cb: F) {
  //   let v = this.vertices[index]
  //   if (v !== undefined) {
  //     let color = this.initColor()
  //     let queue: Q<T> = new Queue()
  //     queue.enqueue(v)
  //     color.set(v, 'grey')
  //     while (queue.size()) {
  //       let u = queue.dequeue()
  //       let neibors = this.adjList.get(u)
  //       neibors.forEach((item) => {
  //         if (color.get(item) === 'white') {
  //           queue.enqueue(item)
  //           color.set(item, 'grey')
  //         }
  //       })
  //       cb && cb(u)
  //       color.set(u, 'black')
  //     }
  //   }
  // }
  // _dfs(v: T, color: Map<T, GC>, cb: F) {
  //   cb && cb(v)
  //   color.set(v, 'grey')
  //   let neibors = this.adjList.get(v)
  //   for (let i = 0; i < neibors.length; i++) {
  //     let w = neibors[i]
  //     if (color.get(w) === 'white') {
  //       this._dfs(w, color, cb)
  //     }
  //   }
  //   color.set(v, 'black')
  // }
  // dfs(index = 0, cb: F) {
  //   let v = this.vertices[index]
  //   if (v !== undefined) {
  //     let color = this.initColor()
  //     this._dfs(v, color, cb)
  //   }
  // }
  // shortestPath(index = 0) {
  //   let distance = new Map()
  //   let predecessors = new Map()
  //   let v = this.vertices[index]
  //   if (v) {
  //     let color = this.initColor()
  //     let queue: Q<T> = new Queue()
  //     queue.enqueue(v)
  //     color.set(v, 'grey')
  //     // init
  //     this.vertices.forEach((item) => {
  //       distance.set(item, 0)
  //       predecessors.set(item, null)
  //     })
  //     while (!queue.size()) {
  //       let u = queue.dequeue()
  //       let neibors = this.adjList.get(u)
  //       neibors.forEach((item) => {
  //         if (color.get(item) === 'white') {
  //           queue.enqueue(item)
  //           color.set(item, 'grey')
  //           distance.set(item, distance.get(u) + 1)
  //           predecessors.set(item, u)
  //         }
  //       })
  //       color.set(u, 'black')
  //     }
  //   }
  //   return {
  //     distance,
  //     predecessors,
  //   }
  // }
  // getPath(fromIndex: N, toIndex: N) {
  //   let queue: Q<T> = new Queue()
  //   let to = this.vertices[toIndex]
  //   let from = this.vertices[fromIndex]
  //   if (to && from) {
  //     let { predecessors } = this.shortestPath(fromIndex)
  //     let cur = to
  //     while (cur) {
  //       queue.enqueue(cur)
  //       cur = predecessors.get(cur)
  //     }
  //   }
  //   queue.reverse()
  //   return queue
  // }
  // neighborsMatrix() {
  // }
  // neighborsTable() {
  //   return this.adjList
  // }
}

export { Graph }
