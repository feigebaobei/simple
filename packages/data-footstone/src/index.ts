import { Stack } from './stack'
import { Queue, PriorityQueue } from './queue'
import {
  SingleChain,
  DoublyChain,
  SingleCircleChain,
  DoublyCircleChain,
} from './chain'
import { PSet } from './pSet'
import { PMap } from './pMap'
import { HashMap, djb2HashFn, loseloseHashFn } from './hashMap'
import { BaseTree, BinarySearchTree, AVLTree, RedBackTree } from './tree'
import * as order from './order'
import { Lru, Fifo } from './store'

export {
  Stack,
  Queue,
  PriorityQueue,
  SingleChain,
  DoublyChain,
  SingleCircleChain,
  DoublyCircleChain,
  PSet,
  PMap,
  HashMap,
  djb2HashFn,
  loseloseHashFn,
  BaseTree,
  BinarySearchTree,
  AVLTree,
  RedBackTree,
  order,
  Lru,
  Fifo,
}
