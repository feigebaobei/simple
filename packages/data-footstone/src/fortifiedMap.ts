import { FortifiedMap as FM, } from '../typings'

class FortifiedMap<T, G> implements FM<T, G> {
    box: Map<T, G>
    constructor () {
        this.box = new Map()
    }
    set(k: T, v: G) {
        this.box.set(k, v)
    }
    delete(k: T) {
        return this.box.delete(k)
    }
    has(k: T) {
        return this.box.has(k)
    }
    get(k: T) {
        return this.box.get(k)
    }
    clear() {
        return this.box = new Map()
    }
    size() {
        return this.box.size
    }
    keys() {
        return this.box.keys()
    }
    values() {
        return this.box.values()
    }
    getMap() {
        return this.box
    }
}

export {FortifiedMap}