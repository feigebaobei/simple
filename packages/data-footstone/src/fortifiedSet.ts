import { FortifiedSet as FS } from '../typings'

class FortifiedSet<T> implements FS<T> {
    box: Set<T>
    constructor () {
        this.box = new Set()
        // this[Symbol.iterator]: a[Symbol.iterator].bind(a),
        this[Symbol.iterator] = this.values()
    }
    // box: Set<T>
    // add: (v: T) => {}
    add(v: T) {
        this.box.add(v)
    }
    delete(v: T) {
        return this.box.delete(v)
    }
    has(v: T) {
        return this.box.has(v)
    }
    clear() {
        return this.box = new Set()
    }
    size() {
        return this.box.size
    }
    values() {
        return this.box.values()
    }
    getSet() {
        return this.box
    }
    // // 并集
    concat(...v: PSet<T>[]) {
        v.forEach(item => {
            item.forEach(subItem => {
                this.add(subItem)
            })
        })
        return this
    }
    // // 交集
    intersect(...v: PSet<T>[]) {
        v.forEach(item => {
            item.forEach(subItem => {
                if (!this.has(subItem)) {
                    this.delete(subItem)
                }
            })
        })
        return this
    }
    // // 差集
    diffSet(...v: PSet<T>[]) {
        v.forEach(item => {
            item.forEach(subItem => {
                if (this.has(subItem)) {
                    this.delete(subItem)
                }
            })
        })
    }
    // // 子集
    subSetOf(v: PSet<T>) {
        return v.every(item => {
            return this.has(item)
        })
    }

}

export {FortifiedSet}