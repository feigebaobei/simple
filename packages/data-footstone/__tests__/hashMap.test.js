import { HashMap, djb2HashFn, loseloseHashFn } from '../src/hashMap'

describe('HashMap', () => {
  it('HashMap', () => {
    let hm = new HashMap()
    expect(hm.kind).toBe('separate')
    expect(hm.hash).toBe('djb2')
    hm.put('a', 1)
    expect(hm.get('a')).toBe(1)
    hm.put('b', 2)
    expect(hm.get('b')).toBe(2)
    expect(hm.remove('a')).toBe(1)
    expect(hm.get('a')).toBeUndefined()
    expect(hm.createNode('a', 1)).toEqual({ key: 'a', value: 1 })
    expect(hm.size()).toBe(1)
    expect(hm.get('c')).toBeUndefined()
    expect(hm.remove('c')).toBeUndefined()
  })
  test('HashMap put', () => {
    // 测试插入相同的key.
    let hm = new HashMap()
    expect(hm.put('a', 'str')).toBe(1)
    expect(hm.get('a')).toBe('str')
    expect(hm.put('a', 'string')).toBe(1)
    expect(hm.get('a')).toBe('string')
    hm = new HashMap('line')
    expect(hm.put('a', 'str')).toBe(1)
    expect(hm.get('a')).toBe('str')
    expect(hm.put('a', 'string')).toBe(1)
    expect(hm.get('a')).toBe('string')
  })
  test('HashMap separate conflict', () => {
    let hm = new HashMap('separate', 'loselose')
    expect(hm.hashFn('Ana')).toBe(13)
    expect(hm.hashFn('Donnie')).toBe(13)
    hm.put('Ana', 1)
    hm.put('Donnie', 2)
    expect(hm.get('Ana')).toBe(1)
    expect(hm.get('Donnie')).toBe(2)
    expect(hm.size()).toBe(2)
    expect(hm.remove('Ana')).toBe(1)
    expect(hm.size()).toBe(1)
    // hashFn
  })
  test('HashMap line conflict', () => {
    let hm = new HashMap('line', 'loselose')
    expect(hm.hashFn('Ana')).toBe(13)
    expect(hm.hashFn('Donnie')).toBe(13)
    hm.put('Ana', 1)
    hm.put('Donnie', 2)
    expect(hm.get('Ana')).toBe(1)
    expect(hm.get('Donnie')).toBe(2)
    expect(hm.size()).toBe(2)
    expect(hm.remove('Donnie')).toBe(2)
    expect(hm.size()).toBe(1)
    // hashFn
  })
})

describe('djb2HashFn', () => {
  test('djb2HashFn', () => {
    expect(djb2HashFn('a')).toMatchSnapshot()
    expect(djb2HashFn('b')).toMatchSnapshot()
  })
  test('djb2HashFn conflict', () => {
    // 记录几个有冲突的值，在测试时使用。
    expect(djb2HashFn('Donnie')).toMatchSnapshot() // 278
    expect(djb2HashFn('Ana')).toMatchSnapshot() // 925
    expect(djb2HashFn('Tyrion')).toMatchSnapshot() // 624
    expect(djb2HashFn('Aaron')).toMatchSnapshot() // 215
  })
})
describe('loseloseHashFn', () => {
  test('loseloseHashFn', () => {
    expect(loseloseHashFn('a')).toMatchSnapshot()
    expect(loseloseHashFn('b')).toMatchSnapshot()
  })
  test('loseloseHashFn conflict', () => {
    // 记录几个有冲突的值，在测试时使用。
    expect(loseloseHashFn('Donnie')).toMatchSnapshot() // 13
    expect(loseloseHashFn('Ana')).toMatchSnapshot() // 13
    expect(loseloseHashFn('Tyrion')).toMatchSnapshot() // 16
    expect(loseloseHashFn('Aaron')).toMatchSnapshot() // 16
  })
})
