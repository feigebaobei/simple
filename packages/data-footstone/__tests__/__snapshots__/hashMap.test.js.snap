// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`djb2HashFn djb2HashFn 1`] = `395`;

exports[`djb2HashFn djb2HashFn 2`] = `396`;

exports[`djb2HashFn djb2HashFn 3`] = `278`;

exports[`djb2HashFn djb2HashFn 4`] = `925`;

exports[`djb2HashFn djb2HashFn 5`] = `624`;

exports[`djb2HashFn djb2HashFn 6`] = `215`;

exports[`djb2HashFn djb2HashFn conflict 1`] = `278`;

exports[`djb2HashFn djb2HashFn conflict 2`] = `925`;

exports[`djb2HashFn djb2HashFn conflict 3`] = `624`;

exports[`djb2HashFn djb2HashFn conflict 4`] = `215`;

exports[`loseloseHashFn loseloseHashFn 1`] = `23`;

exports[`loseloseHashFn loseloseHashFn 2`] = `24`;

exports[`loseloseHashFn loseloseHashFn conflict 1`] = `13`;

exports[`loseloseHashFn loseloseHashFn conflict 2`] = `13`;

exports[`loseloseHashFn loseloseHashFn conflict 3`] = `16`;

exports[`loseloseHashFn loseloseHashFn conflict 4`] = `16`;
